### PR TITLE
Add initial code for get

### PR DESCRIPTION
**Pull request by udealwithit**
Code to send request to N nodes and wait for R replies.
Still need to write code to send updated versions back to nodes

### CODE REVIEW

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 1]
```diff
@@ -81,6 +105,35 @@ defmodule Vnode do
     end
   end
 
+  defp wait_read_response(key, value, index, context, state, current_read, num_read) do
+    if current_read < num_read do
+      receive do
+        {:ok, ^key, {other_value, other_context}, other_index} ->
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:111]: Where is the base case which returns the same key, value, context? and catch all case?

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 2]
```diff
@@ -81,6 +105,35 @@ defmodule Vnode do
...
+          cond do
+            VClock.compare_vclocks(context, other_context) == :after ->
+              Logger.info("vclock received. Result of comparison: After")
+              # Need to send a put request to the node behind
+              wait_read_response(key, value, index, context, state, current_read+1, num_read)
+            VClock.compare_vclocks(context, other_context) == :before ->
+              Logger.info("vclock received. Result of comparison: Before")
+              index_map = Map.get(state.data, key, %{})
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:119]: I think you should get `index` not `key`.

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 3]
```diff
@@ -81,6 +105,35 @@ defmodule Vnode do
...
+              index_map = Map.put(index_map, key, {other_value, other_context})
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:120]: Better to use [pipeline operator](https://elixir-lang.org/getting-started/enumerables-and-streams.html#eager-vs-lazy)
> index_map = 
state.data
|> Map.get(index, %{})
|> Map.put(key, {other_value, other_context})

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 4]
```diff
@@ -81,6 +105,35 @@ defmodule Vnode do
...
+              state = %{state | data: index_map}
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:121]: Current `index_map` is only a map of the `index` not a map of map like `state.data`. You should `Map.get_and_update`

>    {_, index_map} =
      state.data
      |> Map.get_and_update(index, fn index_store ->
        {nil, Map.put(index_store, key, {other_value, other_context})}
      end)

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 5]
```diff
@@ -81,6 +105,35 @@ defmodule Vnode do
...
+              wait_read_response(key, other_value, index, other_context, state, current_read+1, num_read)
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:122]: Why does `current_read` get incremented even it has receive different value?

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 6]
```diff
@@ -81,6 +105,35 @@ defmodule Vnode do
...
+            VClock.compare_vclocks(context, other_context) == :equal ->
+              Logger.info("vclock received. Result of comparison: Equal")
+              wait_read_response(key, value, index, context, state, current_read+1, num_read)
+            VClock.compare_vclocks(context, other_context) == :concurrent ->
+              Logger.info("vclock received. Result of comparison: Concurrent")
+              # Need to send all responses as we have divergent branches
+              wait_read_response(key, value, index, context, state, current_read+1, num_read)
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:129]: Ditto. Why did `current_read` got increased?

File: apps/dynamo/lib/dynamo/vnode.ex [hunk 7]
```diff
@@ -113,6 +166,20 @@ defmodule Vnode do
...
 
+  @doc """
+  Callback for get values from replicas
+  """
+  @impl true
+  def handle_cast({:getrepl, sender, index, key}, state) do
+    Logger.info("returning #{key} value stored in replica")
+
+    index_map = Map.get(state.data, index, %{})
+    value = Map.get(index_map, key, :key_not_found)
```
Comment (sw007chun) [inline apps/dynamo/lib/dynamo/vnode.ex:177]: Use pipeline operator

> value = 
state.data
|> Map.get(index, %{})
|> Map.get(key, :key_not_found)

**Review by sw007chun — CHANGES_REQUESTED** [review]

File: apps/dynamo/lib/dynamo/vnode.ex (hunk 1 above)
Comment (udealwithit) [reply apps/dynamo/lib/dynamo/vnode.ex:111]: We dont need to have same key, value and context, just same key everything else can be checked by comparing the contexts.
I'll write a case for all other cases

File: apps/dynamo/lib/dynamo/vnode.ex (hunk 2 above)
Comment (udealwithit) [reply apps/dynamo/lib/dynamo/vnode.ex:119]: yeah missed it

File: apps/dynamo/lib/dynamo/vnode.ex (hunk 3 above)
Comment (udealwithit) [reply apps/dynamo/lib/dynamo/vnode.ex:120]: ok

File: apps/dynamo/lib/dynamo/vnode.ex (hunk 5 above)
Comment (udealwithit) [reply apps/dynamo/lib/dynamo/vnode.ex:122]: Just wrote it for now so that the code doesnt get stuck

Comment (udealwithit) [reply apps/dynamo/lib/dynamo/vnode.ex:122]: And thinking about it it should increment even if value is different, basically we should return all the available values we get from nodes if their vclocks are concurrent, so it should be fine

**Review by sw007chun — APPROVED** [review]

**Pull request merged**